feat: expose cleanup policy in create statements

cprasad1 opened the pull request:
### Description 
Injects `CLEANUP_POLICY` to `CreateSourceProperties` and `CreateSourceAsProperties`.

The cleanup policy is exposed visa the `getCleanupPolicy()` method in the respective property classes. 
### Testing done 
- unit testing
- manual testing

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

cprasad1 commented on ksqldb-engine/src/test/java/io/confluent/ksql/topic/TopicCreateInjectorTest.java:
```diff
@@ -332,9 +332,9 @@ public void shouldUpdateStatementText() {
         equalTo(
-            "CREATE STREAM X WITH (KAFKA_TOPIC='name', PARTITIONS=1, REPLICAS=1, RETENTION_MS=100) AS SELECT *"
+            "CREATE STREAM X WITH (CLEANUP_POLICY='delete', KAFKA_TOPIC='name', PARTITIONS=1, REPLICAS=1, RETENTION_MS=100) AS SELECT *"
                 + "\nFROM SOURCE SOURCE\n"
-                + "EMIT CHANGES;"));
+                + "EMIT CHANGES"));
```
should we add the cleanup policy to the masked statement?

cprasad1 commented on ksqldb-engine/src/test/java/io/confluent/ksql/topic/TopicCreateInjectorTest.java:
```diff
@@ -532,6 +584,74 @@ public void shouldThrowIfRetentionConfigPresentInCreateTableAs() {
+    );
+
+    // Then:
+    assertThat(
+        e.getMessage(),
+        containsString("Invalid config variable in the WITH clause: CLEANUP_POLICY"));
```
we should be blocking users from setting it manually as well, right?

agavra replied:
yes, I think it would be error-prone to allow people to change this configuration

cprasad1 commented on ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java:
```diff
@@ -103,6 +109,10 @@ private ConfiguredStatement<? extends CreateSource> injectForCreateSource(
     final CreateSource createSource = statement.getStatement();
     final CreateSourceProperties properties = createSource.getProperties();
 
+    if (properties.getCleanupPolicy().isPresent()) {
+      throw new KsqlException(CLEANUP_POLICY_PRESENT_EXCEPTION);
+    }
+
```
blocking setting the cleanup policy here

jnh5y commented on ksqldb-common/src/main/java/io/confluent/ksql/properties/with/CommonCreateConfigs.java:
```diff
@@ -197,6 +198,12 @@ public static void addToConfigDef(
             ConfigDef.Type.INT,
             null,
             Importance.LOW,
+            "Undocumented feature"
```
Undocumented?

Is there a reason for it to be undocumented?

cprasad1 replied:
The initial hunch to keep it undocumented was because we don't want users to set the cleanup policy manually. The policy is inferred by our planner based on the type of source we are creating (stream/table/windowed/non-windowed/C*/C*AS). I have updated the PR and added the config doc copied from kafka

jnh5y commented:
Any upgrade concerns when this feature is rolled out? 

How are old queries handled?

agavra commented on ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java:
```diff
@@ -51,6 +52,11 @@
 
+  private static final String CLEANUP_POLICY_PRESENT_EXCEPTION =
+      String.format(
+          "Invalid config variable in the WITH clause: %s",
```
It'd be good to explain _why_ it's invalid in the error message

cprasad1 replied:
changed to:
```
Invalid config variable in the WITH clause: CLEANUP_POLICY.
The CLEANUP_POLICY config is automatically inferred based on the type of source (STREAM or TABLE).
Users can't set the CLEANUP_POLICY config manually.
```

agavra reviewed (approved):
LGTM! Thanks @cprasad1

cprasad1 commented on ksqldb-engine/src/main/java/io/confluent/ksql/topic/TopicCreateInjector.java:
```diff
@@ -190,17 +204,15 @@ private <T extends CreateAsSelect> ConfiguredStatement<?> injectForCreateAsSelec
-            .getOrDefault(TopicConfig.RETENTION_MS_CONFIG, info.getRetentionInMillis())
-    ));
-
-    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
```
added back

cprasad1 commented:
@jnh5y 
>Any upgrade concerns when this feature is rolled out? How are old queries handled?
The quick and dirty answer: I don't see any risks with old queries since the topologies/plans are not changing. If they change, then we have tests to guard against them afaik. This just exposes the cleanup policy that was already being computed. The logic for the computation of the cleanup policy isn't changing. 

Let me know if you see other risks. I will think about it as well for a bit more.

cprasad1 commented:
addressed all comments. need a final +1 @agavra @jnh5y .

@jnh5y regarding upgrade concerns: The old queries would not even have the new config in their configured statement. This is a config that was already being computed but not exposed via the props earlier. We just expose them now

cprasad1 commented:
green buil https://jenkins.confluent.io/job/Confluent%20Public%20Repo%20PR%20builder/job/ksql/job/PR-9512/8/. Merging
